File: median_array.py
# Time Complexity : O(n log m) where n is the length of longer smaller array and m is the length of smaller array
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Three line explanation of solution in plain english
# I use partitions and find the correct partition using binary search. Once correct partition is found, return middle element or average of 2 middle elements if 
# total element are even

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        m = len(nums1)
        n = len(nums2)
        if m > n: return self.findMedianSortedArrays(nums2, nums1)
        left = 0
        right = len(nums1)
        while left <= right:
            mid = left + (right - left) // 2
            p1 = mid
            p2 = (m+n)//2 - p1
            
            l1 = nums1[p1-1] if p1 > 0 else -math.inf
            l2 = nums2[p2-1] if p2 > 0 else -math.inf
            r1 = nums1[p1] if p1 < len(nums1) else math.inf
            r2 = nums2[p2] if p2 < len(nums2) else math.inf

            if l1 <= r2 and l2 <= r1:
                if (m + n) % 2 == 0:
                    return (max(l1,l2) + min(r1, r2)) / 2
                else:
                    return min(r1, r2)   
            elif l2 > r1:
                left = mid+1
            else:
                right = mid-1
            
        return -1

File: MedianArray.java
/*
worked on LC : YES
Time complexity : O(logn)
spacce complexity : O(1)
*/

class Solution {
    
    // 1 7 8 10 11
    // 2 3 4 5 6 12 13
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        
        
        int n = nums1.length;
        int m  = nums2.length;
        
        if(n > m) {
            return findMedianSortedArrays(nums2, nums1);
        }
        
        int low = 0;
        int high = n;
        
        
        while(low <= high) {
            
            int partX = low + (high - low) /2;
            
            int partY = (m+n)/2 - partX;
            
//             boundy elements and make partitions
            //             there are no element in first array left 
            double l1 =  (partX == 0 ) ? Integer.MIN_VALUE : nums1[partX-1];
            //             there are no element in first array right 
            double r1 =  (partX == n ) ? Integer.MAX_VALUE : nums1[partX];
            //             there are no element in second array left
            double l2 =  (partY == 0 ) ? Integer.MIN_VALUE : nums2[partY-1];
            //             there are no element in second array right
            double r2 =  (partY == m ) ? Integer.MAX_VALUE : nums2[partY];
            
            
            if(l1 <= r2 && l2 <= r1) {
                // found the partition
                // if even 
                if((m + n)%2  == 0 ) {
                    return (Math.max(l1,l2) + Math.min(r1,r2) )/2;
                }else{
                    // if odd
                    return Math.min(r1, r2);
                }
            }else if ( l2 > r1) {
                //  increase left partition                     
                low = partX+1;
            }else{
                high = partX-1;
            }
            
            
        }
        
        return Double.MIN_VALUE;
        
    }
}
